event_model/tests/test_em.py

```diff
@@ -340,6 +340,154 @@ def test_document_router_():
+            assert doc['data']['x'][0] == 1
+            mutable.append(object())
+
+    dr = DefinesEventPageAndEvent()
+    # Test that Event is routed to Event.
+    dr('event', event1)
+    assert len(mutable) == 1
+    mutable.clear()
+    # Test that EventPage is routed to EventPage.
+    dr('event_page', event_page)
+    assert len(mutable) == 1
+    mutable.clear()
+
+
+def test_document_router_dispatch_datum():
+
+    mutable = []  # used for counting calls
+
+    # example documents
+    datum1 = {'datum_id': 'placeholder/1',
+              'resource': 'placeholder',
+              'datum_kwargs': {'index': 1}}
+    datum2 = {'datum_id': 'placholder/2',
+              'resource': 'placeholder',
+              'datum_kwargs': {'index': 2}}
+    datum_page = event_model.pack_datum_page(datum1, datum2)
+
+    class DefinesDatumNotDatumPage(event_model.DocumentRouter):
+        def datum(self, doc):
+            # Just a dumb test that check something particular to these example
+            # documents.
+            assert doc['datum_kwargs']['index'] == int(doc['datum_id'][-1])
+            mutable.append(object())
+
+    dr = DefinesDatumNotDatumPage()
+    # Test that Datum is routed to Datum.
+    dr('datum', datum1)
+    assert len(mutable) == 1
+    mutable.clear()
+    # Test that DatumPage is unpacked and routed to Datum one at a time.
+    dr('datum_page', datum_page)
+    assert len(mutable) == 2
+    mutable.clear()
+
+    class DefinesDatumPageNotDatum(event_model.DocumentRouter):
+        def datum_page(self, doc):
+            # Just a dumb test that check something particular to these example
+            # documents.
+            assert doc['datum_kwargs']['index'][0] == int(doc['datum_id'][0][-1])
+            mutable.append(object())
+
+    dr = DefinesDatumPageNotDatum()
+    # Test that Datum is packed and routed to DatumPage.
+    dr('datum', datum1)
+    assert len(mutable) == 1
+    mutable.clear()
+    # Test that DatumPage is routed to DatumPage.
+    dr('datum_page', datum_page)
+    assert len(mutable) == 1
+    mutable.clear()
+
+    class DefinesDatumPageAndDatum(event_model.DocumentRouter):
+        def datum(self, doc):
+            # Just a dumb test that check something particular to these example
+            # documents.
+            assert doc['datum_kwargs']['index'] == int(doc['datum_id'][-1])
+            mutable.append(object())
+
+        def datum_page(self, doc):
+            # Just a dumb test that check something particular to these example
+            # documents.
+            assert doc['datum_kwargs']['index'][0] == int(doc['datum_id'][0][-1])
+            mutable.append(object())
+
+    dr = DefinesDatumPageAndDatum()
+    # Test that Datum is routed to Datum.
+    dr('datum', datum1)
+    assert len(mutable) == 1
+    mutable.clear()
+    # Test that DatumPage is routed to DatumPage.
+    dr('datum_page', datum_page)
+    assert len(mutable) == 1
+    mutable.clear()
```

Comment:
Same as above. I think that we need two separate mutable lists to prove that the correct method has been called.